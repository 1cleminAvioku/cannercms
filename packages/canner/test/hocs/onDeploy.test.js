

import * as React from 'react';
import {get} from 'lodash';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import withOnDeploy, {getValueAndPaths} from '../../src/hocs/onDeploy';
import RefId from 'canner-ref-id';

Enzyme.configure({ adapter: new Adapter() });

describe('getValueAndPaths', () => {
  it('should works', () => {
    const rootValue = {
      posts: {
        edges: [{
          cursor: 'id1',
          node: {
            id: 'id1',
            images: [{
              url: 'url1'
            }]
          }
        }],
        pageInfo: {
          hasNextInfo: false
        }
      }
    };

    const idPathArr = ['posts', '0', 'images', '0', 'url'];
    expect(getValueAndPaths(get(rootValue, ['posts', 'edges', '0', 'node']), idPathArr.slice(2))).toEqual({
      value: 'url1',
      paths: ['images', '0', 'url']
    });
  });
});

describe('withOnDeploy', () => {
  let WrapperComponent, props, MockComponent, mockOnDeploy, mockRemoveOnDeploy;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }

    mockOnDeploy = jest.fn();
    mockRemoveOnDeploy = jest.fn();

    props = {
      refId: new RefId('posts/0/images/0/url'),
      pattern: 'array.array.string',
      onDeploy: mockOnDeploy,
      removeOnDeploy: mockRemoveOnDeploy,
      rootValue: {
        posts:[{
          id: 'id1',
          images: [{
            url: 'url1'
          }]
        }]
      },
  
    }

    WrapperComponent = withOnDeploy(MockComponent);
  });

  it('should render', () => {
    const wrapper = mount(
        <WrapperComponent
          {...props}
        />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should onDeploy works', () => {
    const wrapper = mount(
      <WrapperComponent
        {...props}
      />
    );
    const callback = jest.fn().mockImplementation(() => {
      return 'new url';
    });

    wrapper.instance().onDeploy(callback);
    expect(mockOnDeploy.mock.calls[0][0]).toBe('posts');
    expect(typeof mockOnDeploy.mock.calls[0][1]).toBe('function');
    expect(mockOnDeploy.mock.calls[0][1]({data: get(props.rootValue, ['posts', 0])})).toMatchSnapshot();
    expect(callback).toHaveBeenCalledTimes(1);
  });
});