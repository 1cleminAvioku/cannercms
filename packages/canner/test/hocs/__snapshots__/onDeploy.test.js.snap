// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withOnDeploy should onDeploy works 1`] = `
Object {
  "data": Object {
    "0": "new url",
    "id": "id1",
    "images": Array [
      Object {
        "url": "url1",
      },
    ],
  },
}
`;

exports[`withOnDeploy should render 1`] = `
<ComponentWithOnDeploy
  onDeploy={[MockFunction]}
  pattern="array.array.string"
  refId={
    CannerRefId {
      "_internalArr": Array [
        "posts",
        "0",
        "images",
        "0",
        "url",
      ],
      "child": [Function],
      "getPathArr": [Function],
      "remove": [Function],
      "toString": [Function],
    }
  }
  removeOnDeploy={[MockFunction]}
  rootValue={
    Object {
      "posts": Array [
        Object {
          "id": "id1",
          "images": Array [
            Object {
              "url": "url1",
            },
          ],
        },
      ],
    }
  }
>
  <MockComponent
    onDeploy={[Function]}
    pattern="array.array.string"
    refId={
      CannerRefId {
        "_internalArr": Array [
          "posts",
          "0",
          "images",
          "0",
          "url",
        ],
        "child": [Function],
        "getPathArr": [Function],
        "remove": [Function],
        "toString": [Function],
      }
    }
    removeOnDeploy={[Function]}
    rootValue={
      Object {
        "posts": Array [
          Object {
            "id": "id1",
            "images": Array [
              Object {
                "url": "url1",
              },
            ],
          },
        ],
      }
    }
  >
    <div>
      Component
    </div>
  </MockComponent>
</ComponentWithOnDeploy>
`;
